Source/CombatExtended/Harmony/Harmony_GenRadial.cs

```diff
@@ -79,54 +79,92 @@ internal static void Patch()
-            var radialPatternNumCells = RadialPatternNumCells; // cache friendly local references
-            var radialPatternRadii = RadialPatternRadii;
-            if (radius >= MAX_RADIUS)
-            {
-                if (radius > MAX_RADIUS)
+                if (radius < 0f)
+                    __result = 0;
+                else if (radius < 1f)
+                    __result = 1;
+                else if (radius < 1.414213562f)
+                    __result = 5;
+                else if (radius < 2f)
+                    __result = 9;
+                else
                 {
-                    Log.Error($"Requested radius {radius} is beyond max. Truncating to {Harmony_GenRadial.MAX_RADIUS}.");
+                    __result = RadialPatternCount;
+                    CheckRadius(radius);
                 }
-                __result = RadialPatternCount;
                 return false;
             }
-            float radsq = radius * radius;
-            int r = (int)radsq;
-            int count = radialPatternNumCells[r - 1];
-            /*
-              If we raise the max radius above about 200, binary search becomes faster.
-              Below 200, the match will reliably be found within 64 tries, where linear memory access dominates.
-            int max_count = RadialPatternNumCells[r];
-            int span;
-            */
+
+            int radiusSquaredInt = (int)(radius * radius);
+            // The answer is at least (25/8 * radius * radius - 11)
+            int lowerBound = ((radiusSquaredInt * 25) >> 3) - 11;
 #if BINSEARCH
-            while ((span = max_count - count) > 64)
+            // The upper bound is set to (101/32 * radius * radius)
+            //   some small radius actually goes above that
+            //   but the error is small (< 16) and linear search can fix it
+            int upperBound = (radiusSquaredInt * 101) >> 5;
+            if (upperBound > RadialPatternCount)
+                upperBound = RadialPatternCount;
+            // If the gap is large, perform binary search
+            while (upperBound - lowerBound > 88)
             {
-                int mid_count = count + span / 2;
-                float mid = radialPatternRadii[mid_count];
-                if (mid > radius)
-                {
-                    max_count = mid_count;
-                }
+                int mid = (lowerBound + upperBound) >> 1;
+                if (RadialPatternRadii[mid] <= radius)
+                    lowerBound = mid;
                 else
-                {
-                    count = mid_count;
-                }
+                    upperBound = mid;
             }
 #endif
-            float start = radialPatternRadii[count];
-            while (start <= radius)
-            {
-                start = radialPatternRadii[count++];
-            }
-            __result = count - 1;
+            // Linear search:
+            // When (int)radius + (int)diagonal is odd,  the answer is always 8n + 5
```

Comment:
You mention `diagonal`, but don't define that. I can see that 8n in this context is going to be due to the 8-fold symmetry of the coords, with  the +5 coming from the on-axis values, but I'm not sure that will be obvious to future maintainers.